### PR TITLE
promtail: Add `max-line-size-truncate`

### PR DESCRIPTION
**Pull request by nicoche**
**What this PR does / why we need it**:

It allows promtail to truncate too long lines instead of just dropping them. See https://github.com/grafana/loki/pull/8153#issuecomment-1384039579 for details

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
- [x] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

**Comment by nicoche** [comment]
@kavirajk cf. our discussion in https://github.com/grafana/loki/pull/8202, let me know what you think 🙂

File: clients/pkg/promtail/client/client.go
Comment (kavirajk) [inline clients/pkg/promtail/client/client.go:42]: We don't need extra label `mutationLabel`. let's use `reason` label itself.

File: clients/pkg/promtail/client/client.go
Comment (kavirajk) [inline clients/pkg/promtail/client/client.go:49]: Let's reuse same reason `line_too_long` for both discarding and truncating long lines (won't be a backward incompatible as we haven't released this yet)

1. rename `DropReasongMaxLineSizeLimited` -> `ReasonLineTooLong` = "line_too_long"
2. remove `MutationTruncated`. We don't need this.

Now use `reason` label as `ReasonLineTooLong` in both truncated and discarded related metrics when max_line_size is exceeded.

File: clients/pkg/promtail/client/client.go
Comment (kavirajk) [inline clients/pkg/promtail/client/client.go:70]: We also don't need this `countersWithHostTenantMutation`. Just reuse `countersWithHostTenantReason`.

File: clients/pkg/promtail/client/client.go
Comment (kavirajk) [inline clients/pkg/promtail/client/client.go:104]:
```suggestion
		Help:      "The total number of log entries that have been mutated.",
```

File: clients/pkg/promtail/client/client.go
Comment (kavirajk) [reply clients/pkg/promtail/client/client.go:42]: Rename `DropReasonLabel` -> just `ReasonLabel`.

File: clients/pkg/promtail/client/client.go
Comment (kavirajk) [inline clients/pkg/promtail/client/client.go:105]:
```suggestion
	}, []string{HostLabel, TenantLabel, ReasonLabel})
```

File: clients/pkg/promtail/client/client.go
Comment (kavirajk) [inline clients/pkg/promtail/client/client.go:101]: Can you add one more metrics to track mutated bytes? similar to `droppedBytes`.

For example.
```
var MutatedBytes = prometheus.NewCounterVec(
	prometheus.CounterOpts{
		Namespace: "promtail",
		Name:      "mutated_bytes_total",
		Help:      "The total number of bytes that have been mutated.",
	},
	[]string{HostLabel, TenantLabel, ReasonLabel}
)

```

**Review by kavirajk — COMMENTED** [review]
Thanks @nicoche for the PR. Overall looks good. Left few comments

**Comment by nicoche** [comment]
Thanks for the feedback! I incorporated your suggestions 🙂

File: clients/pkg/promtail/client/client.go
Comment (kavirajk) [inline clients/pkg/promtail/client/client.go:311]: Here I would prefer to handle early return case first. So that we don't need to have unwanted `if and else` (reads bit better IMO)

```go
	if !c.maxLineSizeTruncate {
		c.metrics.droppedEntries.WithLabelValues(c.cfg.URL.Host, tenantID, ReasonLineTooLong).Inc()
		c.metrics.droppedBytes.WithLabelValues(c.cfg.URL.Host, tenantID, ReasonLineTooLong).Add(len(e.Line))
		break
	}
	c.metrics.mutatedEntries.WithLabelValues(c.cfg.URL.Host, tenantID, ReasonLineTooLong).Inc()
	c.metrics.mutatedBytes.WithLabelValues(c.cfg.URL.Host, tenantID, ReasonLineTooLong).Add(float64(len(e.Line) - c.maxLineSize))
	e.Line = e.Line[:c.maxLineSize]
```

NOTE: I also update `c.metrics.droppedBytes` when skipping the log line completely.

File: clients/pkg/promtail/client/client_test.go
Comment (kavirajk) [inline clients/pkg/promtail/client/client_test.go:101]:
```suggestion
		"dropping log entries that have max_line_size exceeded": {
```

File: clients/pkg/promtail/client/client_test.go
Comment (kavirajk) [inline clients/pkg/promtail/client/client_test.go:139]:
```suggestion
		"truncating log entries that have max_line_size exceeded": {
```

**Review by kavirajk — COMMENTED** [review]
Thanks @nicoche. Almost there :)  Left few minor suggestions. Happy to merge once they are addressed.

File: clients/pkg/promtail/client/client.go
Comment (nicoche) [reply clients/pkg/promtail/client/client.go:309]: Good catch!

File: clients/pkg/promtail/client/client.go
Comment (nicoche) [reply clients/pkg/promtail/client/client.go:66]: Good catch, forgot this one 🙂

**Comment by nicoche** [comment]
Nice catches 🙂. Updated  the PR

File: clients/pkg/promtail/limit/config.go
Comment (kavirajk) [inline clients/pkg/promtail/limit/config.go:16]: `mapstructure:"max_line_size"` -> `mapstructure:"max_line_size_truncate"`

File: docs/sources/clients/promtail/configuration.md
Comment (kavirajk) [inline docs/sources/clients/promtail/configuration.md:1947]: Would prefer to have exact help message we have it on the flag parsing.
`Whether to truncate lines that exceed max_line_size. No effect if max_line_size is disabled`

**Review by kavirajk — COMMENTED** [review]
Thanks @nicoche. LGTM. Two minor changes needed before merge.

 **NOTE**: In general, force push make it hard to know what changed after my last review. Please do normal push after addressing the comments (if needed rebase, do `git merge`)

File: clients/pkg/promtail/limit/config.go
Comment (nicoche) [reply clients/pkg/promtail/limit/config.go]: Good catch!! This one could have easily slipped through.

**Review by kavirajk — APPROVED** [review]
LGTM. Nice work @nicoche :+1:

**Comment by nicoche** [comment]
> LGTM. Nice work @nicoche :+1: 

Thanks for the feedbacks 🙂

**Pull request merged**
